File: src/main/java/plazma/lib/data/yaml/YamlLib.java
/*
 * Copyright (C) 2012-2023 Oleh Hapon ohapon@users.sourceforge.net
 * 
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 * 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307, USA.
 * 
 * Oleh Hapon
 * Kyiv, UKRAINE
 * ohapon@users.sourceforge.net
 */

package plazma.lib.data.yaml;

import java.io.IOException;

import plazma.lib.data.node.Node;

public class YamlLib {
    
    // tokenize

    public static String[] tokenizeYamlFromText(String input) {
        return new YamlTokenizer().tokenizeYamlFromText(input);
    }

    public static String[] tokenizeYamlFromArray(char[] input) {
        return new YamlTokenizer().tokenizeYamlFromArray(input);
    }

    // parse

    public static Node parseYamlFromTokens(String[] tokens) {
        return new YamlParser().parseYamlFromTokens(tokens);
    }

    public static Node parseYamlFromTokens(YamlReaderConfig config, String[] tokens) {
        return new YamlParser().parseYamlFromTokens(config, tokens);
    }

    // read

    public static Node readYamlFromText(String input) {
        return new YamlReader().readYamlFromText(input);
    }

    public static Node readYamlFromText(YamlReaderConfig config, String input) {
        return new YamlReader().readYamlFromText(config, input);
    }

    //

    public static Node readYamlFromArray(char[] input) {
        return new YamlReader().readYamlFromArray(input);
    }

    public static Node readYamlFromArray(YamlReaderConfig config, char[] input) {
        return new YamlReader().readYamlFromArray(config, input);
    }

    //

    public static Node readYamlFromFile(String fileName) throws IOException {
        return new YamlReader().readYamlFromFile(fileName);
    }

    public static Node readYamlFromFile(YamlReaderConfig config, String fileName) throws IOException {
        return new YamlReader().readYamlFromFile(config, fileName);
    }

    // write

    public static String writeYamlToText(Node root) {
        return new YamlWriter().writeYamlToText(root);
    }

    public static String writeYamlToText(YamlWriterConfig config, Node root) {
        return new YamlWriter().writeYamlToText(config, root);
    }
    
    //
    
    public static void writeYamlToFile(String fileName, Node root) throws IOException {
        new YamlWriter().writeYamlToFile(fileName, root);
    }

    public static  void writeYamlToFile(String fileName, YamlWriterConfig config, Node root) throws IOException {
        new YamlWriter().writeYamlToFile(fileName, config, root);
    }

    //

    public static void writeYamlToConsole(Node root) throws IOException {
        new YamlWriter().writeYamlToConsole(root);
    }

    public static void writeYamlToConsole(YamlWriterConfig config, Node root) throws IOException {
        new YamlWriter().writeYamlToConsole(config, root);
    }

    // convert

    public static Node convertYamlToJson(Node node) {
        Yaml2JsonConverter converter = new Yaml2JsonConverter();
        Node result = converter.convert(node);
        //if (converter != null) {
        //    converter.destroy();
        //}        
        return result;
    }

    public static Node convertYamlToXml(Node node) {
        Yaml2XmlConverter converter = new Yaml2XmlConverter();
        Node result = converter.convert(node);
        //if (converter != null) {
        //    converter.destroy();
        //}        
        return result;
    }

}

File: src/main/cpp/plazma/lib/data/yaml/yamllib.cpp
#include <string>

#include "yamllib.h"
#include "yamltokenizer.h"
#include "yamlparser.h"

#include "yamlwriter.h"
#include "Yaml2XmlConverter.h"
#include "Yaml2JsonConverter.h"

#include "plazma/lib/io/iolib.h"
#include "plazma/lib/text/textlib.h"

using namespace ext;

namespace yamllib {

    // tokenize

    StringList* tokenizeYamlFromText(const std::string& input) {
        return yamltokenizer::tokenizeYamlFromArray(input.c_str(), input.length());
    }

    StringList* tokenizeYamlFromArray(const char* input) {
        return yamltokenizer::tokenizeYamlFromArray(input);
    }

    StringList* tokenizeYamlFromArray(const char* input, int size) {
        return yamltokenizer::tokenizeYamlFromArray(input, size);
    }

    StringList* tokenizeYamlFromFile(const std::string& fileName) {
        char *text = iolib::readChars(fileName);
        return tokenizeYamlFromArray(text);
    }

    // parse
    
    node::Node* parseYamlFromTokens(StringList* tokens) {
        return yamlparser::parseYamlFromTokens(tokens);
    }

    node::Node* parseYamlFromTokens(yaml::YamlReaderConfig* config, StringList* tokens) {
        return yamlparser::parseYamlFromTokens(config, tokens);
    }

    // read

    node::Node* readYamlFromText(const std::string& input) {
        StringList* tokens = yamltokenizer::tokenizeYamlFromText(input);
        return yamlparser::parseYamlFromTokens(tokens);
    }

    node::Node* readYamlFromText(yaml::YamlReaderConfig* config, const std::string& input) {
        StringList* tokens = yamltokenizer::tokenizeYamlFromText(input);
        return yamlparser::parseYamlFromTokens(config, tokens);
    }

    //

    node::Node* readYamlFromArray(const char* input) {
        StringList* tokens = yamltokenizer::tokenizeYamlFromArray(input);
        return parseYamlFromTokens(tokens);
    }

    node::Node* readYamlFromArray(yaml::YamlReaderConfig* config, const char* input) {
        StringList* tokens = yamltokenizer::tokenizeYamlFromArray(input);
        return parseYamlFromTokens(config, tokens);
    }

    //

    node::Node* readYamlFromFile(const std::string& fileName) {
        char *text = iolib::readChars(fileName);
        return readYamlFromArray(text);
    }

    node::Node* readYamlFromFile(yaml::YamlReaderConfig* config, const std::string& fileName) {
        char *text = iolib::readChars(fileName);
        return readYamlFromArray(config, text);
    }

    // write

    std::string writeYamlToText(node::Node* root) {
        return yamlwriter::writeYamlToText(root);
    }

    std::string writeYamlToText(yaml::YamlWriterConfig* config, node::Node* root) {
        return yamlwriter::writeYamlToText(config, root);
    }

    // write: console

    void writeYamlToConsole(node::Node* root) {
        yamlwriter::writeYamlToConsole(root);
    }

    void writeYamlToConsole(yaml::YamlWriterConfig* config, node::Node* root) {
        yamlwriter::writeYamlToConsole(config, root);
    }

    // write: file

    void writeYamlToFile(const std::string& fileName, node::Node* root) {
        // TODO: Use yamlwriter::writeYamlToFile(fileName, config, root);
        std::string text = writeYamlToText(root);
        iolib::writeText(fileName, text);
    }

    void writeYamlToFile(const std::string& fileName, yaml::YamlWriterConfig* config, node::Node* root) {
        // TODO: Use yamlwriter::writeYamlToFile(fileName, config, root);
        std::string text = writeYamlToText(config, root);
        iolib::writeText(fileName, text);
    }

    // convert

    node::Node* convertYamlToXml(node::Node* node) {
        yaml::Yaml2XmlConverter* converter = new yaml::Yaml2XmlConverter();
        node::Node* result = converter->convert(node);
        if (converter != nullptr) {
            delete converter;
        }        
        return result;
    }

    node::Node* convertYamlToJson(node::Node* node) {
        yaml::Yaml2JsonConverter* converter = new yaml::Yaml2JsonConverter();
        node::Node* result = converter->convert(node);
        if (converter != nullptr) {
            delete converter;
        }        
        return result;
    }

    //

    std::string getTokensText(StringList* tokens) {
        return yamltokenizer::getTokensText(tokens);
    }    

}